### PR TITLE
[ts 4.7] Support optional variance annotations

### LINKED ISSUE #14442: Support TypeScript 4.7 `in` and `out` variance annotations
### 💻

- [ ] Would you like to work on this feature?

### What problem are you trying to solve?

TypeScript 4.7 Beta ships with `in` and` out` modifiers for type parameters on type declarations.

```ts
type Getter<out T> = () => T;

type Setter<in T> = (value: T) => void;

interface State<in out T> {
    get: () => T;
    set: (value: T) => void;
}
```

TypeScript developers using Babel will need to support this syntax.

These `in` and `out` modifiers can only be placed on type parameter declarations for

* `type` aliases
* `interface` declarations
* `class` declarations
* [~*possibly*~ `class` expressions](https://github.com/microsoft/TypeScript/issues/48618)

It is possible to have `in`, `out`, and `in out`. Duplicates are not allowed, and `in` must always come before `out`.

### Describe the solution you'd like

For the next minor version of Babel, we may want to implement this feature.

### Describe alternatives you've considered

There are alternative compilers, but that would not be ideal for Babel users.

### Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by magic-akari**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #14442 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | New Feature
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR Implement `Optional variance annotations`
- https://github.com/microsoft/TypeScript/pull/48240

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14359"></a>

### CODE REVIEW

File: packages/babel-generator/src/generators/typescript.ts
Comment (JLHwung) [inline packages/babel-generator/src/generators/typescript.ts:37]: Can you add a generator test case to `packages/babel-generator/test/fixtures/typescript`?

**Review by JLHwung — APPROVED** [review]
Can you add some test cases?
```tsx
// with plugins `jsx` and `typescript`
<in T>() => {}
```
It should throw SingleTypeParameterWithoutTrailingComma error.

And also a type-variance-like JSX element:
```tsx
<in T>() => {}</in>
```
It should be parsed successfully.

**Comment by magic-akari** [comment]
Test Added. please do not merge until microsoft/TypeScript#48240 is merged.

**Review by existentialism — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
cc @bradzacher if you have any comment regarding the AST shape

File: packages/babel-parser/src/types.js
Comment (bradzacher) [inline packages/babel-parser/src/types.js:1046]: I'm not 100% sold on using boolean flags as it locks us in with naming but I guess that this is the same thing we did with class property modifiers as well so it's not like we're breaking convention in the AST.

Given the precedence - this is probably the best approach, so LGTM

**Comment by JLHwung** [comment]
@magic-akari I think the AST shape is fine. We can add a `allowedModifiers` param to `tsParseTypeParameter` and handle it accordingly for every `tsParseTypeParameter` usage.

File: packages/babel-generator/src/generators/typescript.ts
Comment (liuxingbaoyu) [reply packages/babel-generator/src/generators/typescript.ts:37]: I'm investigating ci failures in main branch, but I don't know much about ts and jsx.
`<in T>() => {}</in>` will fail when BABEL_8_BREAKING is enabled.
Can someone tell me if this is a problem with the parser or with the test?

**Pull request merged**
